app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/ReceiverDevice.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.localFileTransfer
+
+import android.util.Log
+import android.widget.Toast
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.withContext
+import org.kiwix.kiwixmobile.core.BuildConfig
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.extensions.toast
+import org.kiwix.kiwixmobile.localFileTransfer.WifiDirectManager.Companion.copyToOutputStream
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import java.net.ServerSocket
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once the handshake has successfully taken place, this async-task is used to receive files from
+ * the sender device on the FILE_TRANSFER_PORT port. No. of files to be received (and their names)
+ * are learnt beforehand during the handshake.
+ *
+ * A single Task is used for the entire file transfer (the server socket accepts connections as
+ * many times as the no. of files).
+ */
+internal class ReceiverDevice(private val wifiDirectManager: WifiDirectManager) {
+  private var incomingFileName: String? = null
+  suspend fun receive() = withContext(Dispatchers.IO) {
+    try {
+      ServerSocket(WifiDirectManager.FILE_TRANSFER_PORT).use { serverSocket ->
+        Log.d(TAG, "Server: Socket opened at " + WifiDirectManager.FILE_TRANSFER_PORT)
+        val zimStorageRootPath = wifiDirectManager.zimStorageRootPath
+        val fileItems = wifiDirectManager.getFilesForTransfer()
+        var isTransferErrorFree = true
+        if (BuildConfig.DEBUG) Log.d(TAG, "Expecting " + fileItems.size + " files")
+
+        fileItems.forEachIndexed { fileItemIndex, incomingFileName ->
+          if (isActive) {
```

Comment:
`takeWhile`

Reply:
hmm actually no, I am dumb, it would have to be a sequence for that to work like how I was thinking, yeah isActive is fine or to avoid deeper nesting you can call `yield()`... actually I don't mind going `asSequence` here also. There are a few options

Reply:
@macgills I'm confused. There are only 2 nested blocks.

Reply:
yes but we generally try to avoid nesting as it makes code harder to read and in my head I started to like the form of

```
fileItems.asSequence()
.takeWhile{ isActive }
.forEachIndexed{...}
```

mostly because `.takeWhile{ isActive }` just reads really nicely to me

Reply:
>fileItems.asSequence()
.takeWhile{ isActive }
.forEachIndexed{...}

this is awesome! I must ask how do you think like that?

Reply:
Coding with RxJava for years got me very used to how clearly you can express yourself
```
by()
.using()
.aSequencne()
.ofOperators()
```

Also it is very much the imperative/declarative divide. Don't be too concerned with how the computer will execute your code and just try and say what you want to happen. Exploring all the operators in the kotlin stdlib on List will let you know what you have available in your toolbox also, filter/map are just the beginning